### DISCUSSION TITLE
Valetudo 2023.03.0

### DISCUSSION THREAD

**Discussion by Hypfer** [Hypfer/Valetudo#1760] [Release Discussions]
<div align="center">
    <img src="https://raw.githubusercontent.com/Hypfer/Valetudo/master/assets/logo/valetudo_logo_with_name.svg" width="800" alt="valetudo">
    <p align="center"><h2>2023.03.0</h2></p>
</div>

A pretty standard release with a few features and bugfixes

## Obstacle locations

If you run Valetudo on a robot that reports obstacle locations, you will now be able to see them in your map.

![image](https://user-images.githubusercontent.com/974410/227315896-96ce9761-98cc-4807-bb84-bad82b59b09a.png)

If you zoom in enough, they will also show a label with additional information (if available).

![image](https://user-images.githubusercontent.com/974410/227316007-738e085e-c308-4fc0-ba20-2d67baaa4085.png)

Right now, you probably won't be able to see these as there are 0 robots with a public root that support this.
Once that changes though, this feature will be waiting.

## Consumable status

Consumables are now displayed with a progress bar indicating the remaining life.

![image](https://user-images.githubusercontent.com/974410/227313636-43e6007d-c02d-4f8e-a4d2-2d75622015db.png)

Previously, the reported numbers lacked context and thus left users wondering if _3 days remaining_ is a lot or a little or something in-between. Now that should be clear.

## Updater download progress

Speaking of progress bars, the updater has also been extended to report the download progress.

![image](https://user-images.githubusercontent.com/974410/227314205-c066e7ac-36e0-4ee5-b23a-bfea7ec8d14a.png)

This should be comforting when updating valetudo via a slow network connection.

## API Changes

This release also comes with two **breaking changes**.

The first is in how the `ZoneCleaningCapability` works.
To be consistent with the `MapSegmentationCapability`, the `iterations` parameter is now a top level property:

```json
{
  "action": "clean",
  "zones": [
    ...
  ],
  "iterations": 1
}
```

If you do zone cleaning via the REST API or MQTT interface, you will have to slightly modify your payloads.

<br/>
<br/>

The second **breaking change** is also for consistency reasons.

Previously, the `StatusStateAttribute` `flag` was reported via the `<TOPIC PREFIX>/<IDENTIFIER>/StatusStateAttribute/detail` topic.
As that was the only place where it was called `detail` it is now instead available via `<TOPIC PREFIX>/<IDENTIFIER>/StatusStateAttribute/flag`.

## Bugfixes

1. Multiple race conditions in the MQTT Client and other components have been identified and fixed.
If you had trouble with connecting to your MQTT broker, this should likely fix that.

2. This release of Valetudo can now properly handle multiple comma-separated errors reported by some Dreame firmwares.

3. Lastly, the `OperationModeControl` on Viomi has been fixed. If you had trouble getting your robot to mop, that should now work again.

## Misc

By buying more robots, another supported CRL200S variant was found: The `Commodore CVR 200`

![image](https://user-images.githubusercontent.com/974410/227314598-f802175b-f4f4-473e-8968-9e8ec06e4b4b.png)

I also bought a `Xiaomi Robot Vacuum-Mop 2S` and a `Kärcher RCV 3`, however unfortunately those aren't based on the CRL200S but the CRL200SD which is entirely different hardware and hence incompatible with Valetudo right now.

Robots that _are_ based on the supported CRL200S can now select between `1` and `2` iterations for segments and zones starting with this release.

## What's next?

After reevaluating a few things, we have a suspicion that the `Roborock S7 Pro Ultra` might be worth further investigation. 
However, doing that unfortunately will cost at least ~850€.

That alone would be expensive but doable, but it is of course not the only robot we're looking at.
I've been buying quite a few other robots for evaluation including most of the new Dreame lineup, which makes this a very expensive hobby.

While there's nothing to announce yet, what I can say is that we're quite busy looking at interesting robots.

Therefore and as always:
If you want to see Valetudo on more robots and/or like this release, you might want to consider donating:

[https://github.com/sponsors/Hypfer](https://github.com/sponsors/Hypfer)

[https://builder.dontvacuum.me/donations.txt](https://builder.dontvacuum.me/donations.txt)

### Breaking Changes

- **mqtt**: Rename status detail to status flag for consistency [`eb67fca`](http://github.com/Hypfer/Valetudo/commit/eb67fcadeda79e9b3717e546eb793891dc6e9768)
- Move zone iterations from zone-level to command-level to be consistent with segments [`ec449d0`](http://github.com/Hypfer/Valetudo/commit/ec449d0d8e3f8ec79ea38e2f0db6d605d8c534f5)

### Features

- **vendor.dreame**: R2240 support [`aaa01bf`](http://github.com/Hypfer/Valetudo/commit/aaa01bf3c0873e24c1c823b167b907d641b036f3)
- **updater**: Visualize download progress [`992c38f`](http://github.com/Hypfer/Valetudo/commit/992c38f63040b09a74d8a29a52d6aba268538d60)
- **ui**: Consumable status progress bars [`a427045`](http://github.com/Hypfer/Valetudo/commit/a427045c3ad26fc9f98737b6d7b8dc86ca796973)
- Obstacle locations [`222ba01`](http://github.com/Hypfer/Valetudo/commit/222ba014300e4d7701b6639d8a944a9b7ae3a628)
- **mqtt**: Allow ignoring certificate errors for brokers with self-signed certificates [`e99bb6b`](http://github.com/Hypfer/Valetudo/commit/e99bb6b4042a793b85c1a668c701ffd1d85630b2)
- **vendor.viomi**: Iterations via set_repeat [`72e6b84`](http://github.com/Hypfer/Valetudo/commit/72e6b84f070be8b7e90b7987d8ac250d4d0b4618)

### Fixes

- **vendor.dreame**: Handle invalid or unknown status data on the 1C [`931ab79`](http://github.com/Hypfer/Valetudo/commit/931ab7966f6808751c5c19cf839a898aa186fab4)
- **miio**: Continue to poll the map in short intervals for a while after state transition from active to inactive [`537f7af`](http://github.com/Hypfer/Valetudo/commit/537f7af9203257c3f44125c27669c25386d6722a)
- **vendor.roborock**: Remove incorrect mop pattern quirk option [`c261d75`](http://github.com/Hypfer/Valetudo/commit/c261d755bf7e4a84698ec0e1fec70cbaec2cdb33)
- **mqtt**: Trying to exterminate some concurrency gremlins [`7bd9122`](http://github.com/Hypfer/Valetudo/commit/7bd91225dd41fb5b5c7be7cb4a629c5977ab7f74)
- **ui**: Fix zone integration help [`5744e4e`](http://github.com/Hypfer/Valetudo/commit/5744e4ea3afa9ebbafa79c7c737a01904cad23a2)
- **vendor.viomi**: Fix operation mode selection [`6b2a268`](http://github.com/Hypfer/Valetudo/commit/6b2a2686b40113caf01f39037f77dca62dcdb120)
- **networkadvertisement**: Ensure that there's only one scheduled networkStateCheck [`0af4703`](http://github.com/Hypfer/Valetudo/commit/0af470342c59e4fbaf8757678a4c4c99557c9f3a)
- **vendor.dreame**: Gracefully handle comma-separated error codes [`460c5d1`](http://github.com/Hypfer/Valetudo/commit/460c5d100673e222075ef5261321215cec0c8159)
- **vendor.roborock**: Handle fan speed miot updates with correct and incorrect piid [`51a412d`](http://github.com/Hypfer/Valetudo/commit/51a412d120a5f2c15235bebf23a6e2520f42b7f0)

### Refactoring

- **ui**: Cleaner use of mutexes in context of async functions [`1055171`](http://github.com/Hypfer/Valetudo/commit/1055171d24b162121bb40457dc5d7faa2abfc3ed)

### Chores

- **release**: 2023.03.0 [`fc6b905`](http://github.com/Hypfer/Valetudo/commit/fc6b905a99f889473f375eb011bd2d3d2b2599ea)
- Un-bump backend axios version [`b4c9e32`](http://github.com/Hypfer/Valetudo/commit/b4c9e3218711c069334d9d3a3fcbf71b4ae8e249)
- Bump dependencies [`979a8b6`](http://github.com/Hypfer/Valetudo/commit/979a8b6a479bf6b29dbecc959d1ba988d1b2878e)
- **vendor.dreame**: Remove old logger calls used for early reverse engineering [`e2a8833`](http://github.com/Hypfer/Valetudo/commit/e2a8833a824159cae4d5057b756f35ddcd68890e)
- Remove config migration code [`d0172f1`](http://github.com/Hypfer/Valetudo/commit/d0172f1ae5cc4d2c5aaee608df181d68dad6ed34)
- Minor cleanup [`d8d7f82`](http://github.com/Hypfer/Valetudo/commit/d8d7f8206f3b6d70560f0f7c4d17225381fd6741)
- Cleanup [`30f629e`](http://github.com/Hypfer/Valetudo/commit/30f629e578885ce72b2cbba1c6578b556794a416)
- Viomi voicepack checksums do matter [`87eb6af`](http://github.com/Hypfer/Valetudo/commit/87eb6afabb753bc1249e746dee2cb702f16d2828)

<hr /><em>This discussion was created from the release <a href='https://github.com/Hypfer/Valetudo/releases/tag/2023.03.0'>Valetudo 2023.03.0</a>.</em>

**Comment by TheDiddel** [comment]
Hey, since the last update (2023.02.0) the mode can no longer be set - "vacuum" is the only mode that works. I get the following error message: PresetSelectionCapabilityRouter: Received error from robot while handling route "/preset" { body: { name: 'vacuum_and_mop' }, message: 'Error code -1' }
I have already tried the new Release 2023.03.0  but unfortunately this does not change anything.
System information:
Dreame L10 Pro, Firmware Version 1105 Release: 2023.03.0
Mike

**Reply by Hypfer** [reply]
Have you tried upgrading your firmware to 1138?
Was it ever working with 1105?

**Reply by TheDiddel** [reply]
I will try this...thanks - and yes - it was working. 
By the way: Thank you very much for your work and your Support!

**Reply by TheDiddel** [reply]
Upgrading the firmware solved the problem - thanks!!
